samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -161,72 +119,96 @@ class KafkaSystemProducer(systemName: String,
+      //   2b. there are no messages in flight because the producer is closed
+
+      // Flush can be called concurrently for different sources, so any modification of shared objects must be threadsafe
+
+      // We must check for an exception AFTER flush() because when flush() returns all callbacks for messages sent
+      // in that flush() are guaranteed to have completed and we update the exception in the callback.
+      // If there is an exception, we rethrow it here to prevent the checkpoint.
+      val exception = fatalException.get()
+      if (exception != null) {
+        metrics.flushFailed.inc
+        throw new SystemProducerException("Flush failed. One or more batches of messages were not sent!", exception)
+      }
+      trace("Flushed %s." format source)
+    }
+  }
 
-        //if there is an exception thrown from the previous callbacks just before flush, we have to fail the container
-        if (sourceData.exceptionInCallback.get() != null) {
-          metrics.flushFailed.inc
-          throw sourceData.exceptionInCallback.get()
-        } else {
-          trace("Flushed %s." format (source))
+
+  private def handleSendException(currentProducer: Producer[Array[Byte], Array[Byte]], producerException: SystemProducerException, isFatalException: Boolean) = {
+    metrics.sendFailed.inc
+    error(producerException)
+    // The SystemProducer contract is synchronous, so there's no way for us to guarantee that an exception will
+    // be handled by the Task before we recreate the producer, and if it isn't handled, a concurrent send() from another
+    // source could send on the new producer before handling its own exceptions and produce out of order messages.
+    // So we have to handle it right here in the SystemProducer or never recreate the producer.
+    if (dropProducerExceptions) {
+      warn("Ignoring producer exception. All messages in the failed produce request will be dropped!", producerException)
+
+      if (isFatalException) {
+        producerLock.synchronized {
+          // Prevent each callback from recreating producer for the same failed event.
+          if (currentProducer == producer) {
+            info("Creating a new producer for system %s." format systemName)
+            producer = getProducer()
+          }
         }
       }
+    } else {
+      // If there is an exception in the callback, it means that the Kafka producer has exhausted the max-retries
+      // Close producer to ensure messages queued in-flight are not sent and hence, avoid re-ordering
+      // This works because the callback thread is singular and no sends can complete until the callback returns.
+      if (isFatalException) {
+        producer.close(0, TimeUnit.MILLISECONDS)
+        fatalException.compareAndSet(null, producerException)
+      }
+    }
+  }
+
+  private def isFatalException(exception: Exception): Boolean = {
```

Comment:
SerializationException is only thrown synchronously on producer.send. I think IllegalState and ProducerClosed exceptions should also be considered fatal exceptions here. If that's true, it'll be cleaner to special case the catch for SerializationException, and remove this method and the isFatalException parameter from handleSendException.